### PR TITLE
Add dockerfile

### PR DESCRIPTION
**Pull request by dtrifiro**
- Add a `Dockerfile` definition: this copies the built wheel into the container and installs it
- Add a github actions workflow to build the dockerfile

### CODE REVIEW

**Review by markstur — COMMENTED** [review]
This is wonderful.  Thank you @dtrifiro !!

**Review by gkumbhat — COMMENTED** [review]
Hi @dtrifiro , Thanks for adding the dockerfile and updates to the readme. The "official" version of caikit image is getting built using [this](https://github.com/opendatahub-io/caikit-tgis-serving/blob/main/Dockerfile) Dockerfile, which is maintained by RH teams. From the perspective of reducing duplicity and maintenance overhead, what is the plan from OpenData hub perspective for the Dockerfile over there vs building the image in this repo ?

**Comment by Xaenalt** [comment]
So, the overall thinking here is that it's better to have stuff in the upstream, this would let any users find it easier and work with it easier, and so the ODH ones would migrate upstream and we'd just be pulling it down for the RHODS builds. This would also let the upstream build have rolling dependencies, whereas they could be fixed downstream on a release cadence

I think an ideal path is to do more of the development here in caikit's upstream, so that we're not carrying large downstream deltas, what do you think @gkumbhat ?

**Comment by dtrifiro** [comment]
Hey @gkumbhat sorry to always bug you :sweat_smile:, but is there anything I can do to help get this merged?

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+USER caikit
+
+ENV RUNTIME_LIBRARY=caikit_nlp
+# Optional: use `CONFIG_FILES` and the /caikit/ volume to explicitly provide a configuration file and models
+# ENV CONFIG_FILES=/caikit/caikit.yml
+VOLUME ["/caikit/"]
+WORKDIR /caikit
+
+CMD ["python",  "-m", "caikit.runtime"]
```
Comment (gkumbhat) [inline Dockerfile:43]: Since executing through runtime isn't main purpose of this repo, lets remove `caikit.runtime` as the entry point here.

Comment (dtrifiro) [reply Dockerfile:43]: Anything else you propose to use as `CMD` or should we just leave this empty, making `bash` the default command?

File: README.md [hunk 2]
```diff
@@ -18,8 +18,122 @@ Capabilities provided by `caikit-nlp`:
...
+  "input_token_count": 10,
+  "seed": null
+}
+```
+
+All the available API endpoints and protos can be dumped using [`scripts/dump_apis.sh`](/scripts/dump_apis.sh).
+
+#### Docker
+
+`caikit-nlp` can also be deployed as a Docker container. To build the docker image:
```
Comment (gkumbhat) [inline README.md:108]: as per my suggestion above regarding entrypoint, we would also need to modify this section a bit.

File: Dockerfile
Comment (gkumbhat) [inline Dockerfile:1]: Can we add license, description etc to this dockerfile that way a user have those information in the container automatically

**Review by gkumbhat — CHANGES_REQUESTED** [review]

File: README.md (hunk 2 above)
Comment (dtrifiro) [reply README.md]: Done

File: Dockerfile
Comment (dtrifiro) [reply Dockerfile:1]: Added `LICENSE` and `README.md`.

We also have those for free by installing the wheel package:

```console
$ ls -/opt/caikit/lib/python3.9/site-packages/caikit_tgis_backend-0.1.21.dist-info:
INSTALLER  LICENSE  METADATA  RECORD  WHEEL
```
`METADATA` contains  package metadata and the embedded readme

File: README.md [hunk 3]
```diff
@@ -18,8 +18,122 @@ Capabilities provided by `caikit-nlp`:
...
+```
+
+Then convert it:
+
+```python
+import caikit_nlp
+
+model_path="flan-t5-small"
+model = caikit_nlp.text_generation.TextGeneration.bootstrap(model_path)
+model.save(f"{model_path}-caikit")
```
[the remark this answers is not in the document]
Comment (dtrifiro) [reply README.md]: Done

File: README.md [hunk 4]
```diff
@@ -16,20 +16,142 @@ Capabilities provided by `caikit-nlp`:
...
+### Bootstrapping models
+
+`caikit_nlp` can use  Hugging Face models, allowing for direct download and bootstrapping.
+
+For example, to use [google/flan-t5-small](https://huggingface.co/google/flan-t5-small):
+
+```python
+import os
+# The env var ALLOW_DOWNLOADS has to be set to allow model downloads before importing caikit_nlp
+os.environ['ALLOW_DOWNLOADS'] = "1"
```
Comment (gkumbhat) [inline README.md:44]: there are some other ways to configure this as well, like via config file, but I think keeping this referring to environment variable is more helpful that way a user instantly knows that they can just configure this via environment variable. No change needed, but just mentioning it here.

File: README.md [hunk 5]
```diff
@@ -16,20 +16,142 @@ Capabilities provided by `caikit-nlp`:
...
+  "generated_tokens": 5,
+  "finish_reason": "MAX_TOKENS",
+  "producer_id": {
+    "name": "Text Generation",
+    "version": "0.1.0"
+  },
+  "input_token_count": 10,
+  "seed": null
+}
+```
```
Comment (gkumbhat) [inline README.md:98]: Later on we can also move these into the `examples` folder.. We can have 2 parts of the "examples", 1 related to notebooks etc and other related to servers including examples from here 🤔

**Review by gkumbhat — APPROVED** [review]
LGTM. Thank you

**Pull request merged**